### PR TITLE
enhancements for sle15 diskful CN deployment

### PR DESCRIPTION
**Pull request by bybai**
### The PR is for task https://github.com/xcat2/xcat-core/issues/5614  and https://github.com/xcat2/xcat-core/issues/5615 and to implement feature deploy sle15 x86_64 CN

_##Feature description##_
1. Sle15 osname is "sle", sles12 osname is "sles", `copycds` find templates based on osname, in order to support both sles12 and sle15, create link `/opt/xcat/share/xcat/install/sle -> /opt/xcat/share/xcat/install/sles` and `/opt/xcat/share/xcat/netboot/sle -> /opt/xcat/share/xcat/netboot/sles`

2. remove HPC related software module. 

### The UT result
1. copycds x86_64 UT:
```
[root@bybc0602 xcat-core]# rpm -Uvh xCAT-server-2.14.4-snap201809060234.noarch.rpm --nodeps
Preparing...                          ################################# [100%]
Updating / installing...
   1:xCAT-server-4:2.14.4-snap20180906################################# [ 50%]
Cleaning up / removing...
   2:xCAT-server-4:2.14.4-snap20180829################################# [100%]

[root@bybc0602 xcat-core]# ls -l /opt/xcat/share/xcat/netboot/ |grep sle
lrwxrwxrwx 1 root root   33 Sep  6 02:39 sle -> /opt/xcat/share/xcat/netboot/sles
drwxr-xr-x 3 root root 4096 Sep  6 02:38 sles
[root@bybc0602 xcat-core]# ls -l /opt/xcat/share/xcat/install/ |grep sle
lrwxrwxrwx 1 root root   33 Sep  6 02:38 sle -> /opt/xcat/share/xcat/install/sles
drwxr-xr-x 2 root root 4096 Sep  6 02:38 sles

[root@bybc0602 /]# copycds /mnt/xcat/iso/suse/15/x86_64/SLE-15-Installer-DVD-x86_64-GM-DVD1.iso
Copying media to /install/sle15/x86_64/1
Media copy operation successful

[root@bybc0602 /]# lsdef -t osimage |grep sle15
sle15-x86_64-install-compute  (osimage)
sle15-x86_64-netboot-compute  (osimage)
sle15-x86_64-statelite-compute  (osimage)
```
2. copycds ppc64le UT:
```
ppc64le # copycds SLE-15-Packages-ppc64le-GM-DVD1.iso
Copying media to /install/sle15/ppc64le/2
Media copy operation successful
ppc64le # lsdef -t osimage
sle15-ppc64le-install-compute  (osimage)
sle15-ppc64le-netboot-compute  (osimage)
sle15-ppc64le-statelite-compute  (osimage)
```
3. Used sles12.3 x86_64 MN provision sle15 x86_64 CN, the basic OS can be installed.

### CODE REVIEW

**Comment by bybai** [comment]
4. commit 3025787 and 8c828d9 are for https://github.com/xcat2/xcat2-task-management/issues/293,  create zypper repos for all products in sle15 Package ISO. The UT:

a. After `nodeset <cn> osimage=<diskful_osimage>` command, generate autoinst file for compute node, the add_on product entries are as following:
```
        <add_on_products config:type="list">
            <listentry><media_url>http://XCATNEXTSERVERHOOK/install/sle15/ppc64le/2</media_url><product>sle-module-basesystem</product><product_dir>/Module-Basesystem</product_dir></listentry><listentry><media_url>http://XCATNEXTSERVERHOOK/install/sle15/ppc64le/2</media_url><product>sle-module-containers</product><product_dir>/Module-Containers</product_dir></listentry><listentry><media_url>http://XCATNEXTSERVERHOOK/install/sle15/ppc64le/2</media_url><product>sle-module-desktop-applications</product><product_dir>/Module-Desktop-Applications</product_dir></listentry><listentry><media_url>http://XCATNEXTSERVERHOOK/install/sle15/ppc64le/2</media_url><product>sle-module-development-tools</product><product_dir>/Module-Development-Tools</product_dir></listentry><listentry><media_url>http://XCATNEXTSERVERHOOK/install/sle15/ppc64le/2</media_url><product>sle-module-legacy</product><product_dir>/Module-Legacy</product_dir></listentry><listentry><media_url>http://XCATNEXTSERVERHOOK/install/sle15/ppc64le/2</media_url><product>sle-module-live-patching</product><product_dir>/Module-Live-Patching</product_dir></listentry><listentry><media_url>http://XCATNEXTSERVERHOOK/install/sle15/ppc64le/2</media_url><product>sle-module-public-cloud</product><product_dir>/Module-Public-Cloud</product_dir></listentry><listentry><media_url>http://XCATNEXTSERVERHOOK/install/sle15/ppc64le/2</media_url><product>sle-module-sap-applications</product><product_dir>/Module-SAP-Applications</product_dir></listentry><listentry><media_url>http://XCATNEXTSERVERHOOK/install/sle15/ppc64le/2</media_url><product>sle-module-server-applications</product><product_dir>/Module-Server-Applications</product_dir></listentry><listentry><media_url>http://XCATNEXTSERVERHOOK/install/sle15/ppc64le/2</media_url><product>sle-module-web-scripting</product><product_dir>/Module-Web-Scripting</product_dir></listentry><listentry><media_url>http://XCATNEXTSERVERHOOK/install/sle15/ppc64le/2</media_url><product>sle-module-web-scripting</product><product_dir>/Product-HA</product_dir></listentry><listentry><media_url>http://XCATNEXTSERVERHOOK/install/sle15/ppc64le/2</media_url><product>sle-module-web-scripting</product><product_dir>/Product-SLES</product_dir></listentry><listentry><media_url>http://XCATNEXTSERVERHOOK/install/sle15/ppc64le/2</media_url><product>sle-module-web-scripting</product><product_dir>/Product-SLES_SAP</product_dir></listentry>
        </add_on_products>
```

b. The diskful sle15 deployment process have no error, after the installation is complete, the zypper repo files in compute node are created as designed as following:

```
:/etc/zypp/repos.d # ls
Basesystem-Module-15-0.repo            SAP-Applications-Module-15-0.repo
Containers-Module-15-0.repo            Server-Applications-Module-15-0.repo
Desktop-Applications-Module-15-0.repo  SLE-15-SAP-15-0.repo
Development-Tools-Module-15-0.repo     SLEHA15-15-0.repo
Legacy-Module-15-0.repo                SLES15-15-0_0.repo
Live-Patching-15-0.repo                SLES15-15-0.repo
Public-Cloud-Module-15-0.repo          Web-Scripting-Module-15-0.repo
:/etc/zypp/repos.d # more /etc/*release
NAME="SLES"
VERSION="15"
VERSION_ID="15"
PRETTY_NAME="SUSE Linux Enterprise Server 15"
ID="sles"
ID_LIKE="suse"
ANSI_COLOR="0;32"
CPE_NAME="cpe:/o:suse:sles:15"
```

**Comment by bybai** [comment]
5. commit f62183e is for https://github.com/xcat2/xcat-core/issues/5630, `ping` is needed in xcat scripts, it is not installed by default, this cause some xcat scripts failed. so I added `iputils` in sle15 pkglist. 

UT:

a. from xcat.log:
```
+ echo 'Thu Sep 13 07:30:55 EDT 2018 [info]: xcat.deployment: finished firstboot preparation, sending request to 10.5.106.5:3002 for changing status...'
Thu Sep 13 07:30:55 EDT 2018 [info]: xcat.deployment: finished firstboot preparation, sending request to 10.5.106.5:3002 for changing status...
+ updateflag.awk 10.5.106.5 3002
+ cd /
+ [[ sle15 =~ sles ]]
+ '[' 1 = 1 ']'
+ set +x
Thu Sep 13 07:31:09 EDT 2018 [info]: xcat.xcatinstallpost: Running /xcatpost/mypostscript.post
Thu Sep 13 07:31:09 EDT 2018 [info]: xcat.mypostscript: Running postbootscript: otherpkgs
Thu Sep 13 07:31:09 EDT 2018 [info]: xcat.mypostscript: postbootscript otherpkgs return with 0
Thu Sep 13 07:31:09 EDT 2018 [debug]: xcat.mypostscript: node booted, reporting status...
Thu Sep 13 07:31:09 EDT 2018 [info]: xcat.mypostscript: provision completed.(bybc0609)
Thu Sep 13 07:31:09 EDT 2018 [info]: xcat.xcatinstallpost: /xcatpost/mypostscript.post return
```
b. status in node definition:
```
:~ # lsdef bybc0609 -i status
Object name: bybc0609
    status=booted
```

**Comment by bybai** [comment]
6. commit a2162b9 for ospkgs enhancement for sle15

a. updatenode default mode on x86_64:
```
bybc0605:/install/postscripts # updatenode bybc0609 ospkgs
bybc0609: trying to download postscripts...
bybc0609: postscripts downloaded successfully
bybc0609: trying to get mypostscript from 10.5.106.5...
bybc0609: Running postscript: ospkgs
bybc0609: postscript: ospkgs exited with code 0
bybc0609: Running of postscripts has completed.
```

b. updatenode verbose mode on ppc64le:
```
c910f03c09k12:/install/postscripts # updatenode c910f03c09k14 ospkgs -V
[c910f03c09k12]: Running command on c910f03c09k12: ip -4 --oneline addr show |awk -F ' ' '{print $4}'|awk -F '/' '{print $1}' 2>&1

[c910f03c09k12]: Running command on c910f03c09k12: chmod -R a+r /install/postscripts 2>&1

[c910f03c09k12]:   c910f03c09k12: Internal call command: xdsh c910f03c09k14 --nodestatus -s -v -e /install/postscripts/xcatdsklspost 1 -m 10.3.9.12 'ospkgs' --tftp /tftpboot --installdir /install --nfsv4 no -c -V
[c910f03c09k12]: Running command on c910f03c09k12: ip -4 --oneline addr show |awk -F ' ' '{print $4}'|awk -F '/' '{print $1}' 2>&1
[c910f03c09k12]: Running command on c910f03c09k12: hostname 2>&1
[c910f03c09k12]: Running command on c910f03c09k12: /opt/xcat/bin/pping c910f03c09k14 2>&1
[c910f03c09k12]: c910f03c09k14: vpdupdate is not supported on the pSeries KVM Guest platform
[c910f03c09k12]: c910f03c09k14: updating VPD database
[c910f03c09k12]: c910f03c09k14: Running /tmp/15rsOf1VIo.dsh 1 -m 10.3.9.12 ospkgs --tftp /tftpboot --installdir /install --nfsv4 no -c -V
[c910f03c09k12]: c910f03c09k14: trying to download postscripts...
[c910f03c09k12]: c910f03c09k14: trying to download postscripts from http://10.3.9.12/install/postscripts/
[c910f03c09k12]: c910f03c09k14: postscripts are downloaded from 10.3.9.12 successfully.
c910f03c09k14: postscripts downloaded successfully
[c910f03c09k12]: c910f03c09k14: trying to get mypostscript from 10.3.9.12...
[c910f03c09k12]: c910f03c09k14: trying to download http://10.3.9.12/tftpboot/mypostscripts/mypostscript.c910f03c09k14...
[c910f03c09k12]: c910f03c09k14: mypostscript.c910f03c09k14 is downloaded successfully.
[c910f03c09k12]: c910f03c09k14: Running //xcatpost/mypostscript
[c910f03c09k12]: c910f03c09k14: Running postscript: ospkgs
[c910f03c09k12]: c910f03c09k14: NFSSERVER=10.3.9.12
c910f03c09k14: OSPKGDIR=/install/sle15/ppc64le/1
c910f03c09k14: OSPKGS = iputils,openssl,rsync
[c910f03c09k12]: c910f03c09k14: pkgs= iputils openssl rsync
c910f03c09k14: groups=
c910f03c09k14: remove groups=
c910f03c09k14: remove pkgs=
[c910f03c09k12]: c910f03c09k14: zypper --non-interactive refresh
c910f03c09k14: Repository 'Basesystem-Module 15-0' is up to date. Repository 'Containers-Module 15-0' is up to date. Repository 'Desktop-Applications-Module 15-0' is up to date. Repository 'Development-Tools-Module 15-0' is up to date. Repository 'Legacy-Module 15-0' is up to date. Repository 'Live-Patching 15-0' is up to date. Repository 'Public-Cloud-Module 15-0' is up to date. Repository 'SAP-Applications-Module 15-0' is up to date. Repository 'SLE-15-SAP 15-0' is up to date. Repository 'SLEHA15 15-0' is up to date. Repository 'SLES15-15-0' is up to date. Repository 'SLES15 15-0' is up to date. Repository 'Server-Applications-Module 15-0' is up to date. Repository 'Web-Scripting-Module 15-0' is up to date. All repositories have been refreshed.
[c910f03c09k12]: c910f03c09k14: zypper --non-interactive refresh
[c910f03c09k12]: c910f03c09k14: Repository 'Basesystem-Module 15-0' is up to date. Repository 'Containers-Module 15-0' is up to date. Repository 'Desktop-Applications-Module 15-0' is up to date. Repository 'Development-Tools-Module 15-0' is up to date. Repository 'Legacy-Module 15-0' is up to date. Repository 'Live-Patching 15-0' is up to date. Repository 'Public-Cloud-Module 15-0' is up to date. Repository 'SAP-Applications-Module 15-0' is up to date. Repository 'SLE-15-SAP 15-0' is up to date. Repository 'SLEHA15 15-0' is up to date. Repository 'SLES15-15-0' is up to date. Repository 'SLES15 15-0' is up to date. Repository 'Server-Applications-Module 15-0' is up to date. Repository 'Web-Scripting-Module 15-0' is up to date. Retrieving repository 'xCAT-sle15-path0' metadata [.done] Building repository 'xCAT-sle15-path0' cache [....done] Retrieving repository 'xCAT-sle15-path1' metadata [.done] Building repository 'xCAT-sle15-path1' cache [....done] Retrieving repository 'xCAT-sle15-path10' metadata [.done] Building repository 'xCAT-sle15-path10' cache [....done] Retrieving repository 'xCAT-sle15-path11' metadata [.done] Building repository 'xCAT-sle15-path11' cache [....done] Retrieving repository 'xCAT-sle15-path12' metadata [.done] Building repository 'xCAT-sle15-path12' cache [....done] Retrieving repository 'xCAT-sle15-path13' metadata [.done] Building repository 'xCAT-sle15-path13' cache [....done] Retrieving repository 'xCAT-sle15-path2' metadata [.done] Building repository 'xCAT-sle15-path2' cache [....done] Retrieving repository 'xCAT-sle15-path3' metadata [.done] Building repository 'xCAT-sle15-path3' cache [....done] Retrieving repository 'xCAT-sle15-path4' metadata [.done] Building repository 'xCAT-sle15-path4' cache [....done] Retrieving repository 'xCAT-sle15-path5' metadata [.done] Building repository 'xCAT-sle15-path5' cache [....done] Retrieving repository 'xCAT-sle15-path6' metadata [.done] Building repository 'xCAT-sle15-path6' cache [....done] Retrieving repository 'xCAT-sle15-path7' metadata [.done] Building repository 'xCAT-sle15-path7' cache [....done] Retrieving repository 'xCAT-sle15-path8' metadata [.done] Building repository 'xCAT-sle15-path8' cache [....done] Retrieving repository 'xCAT-sle15-path9' metadata [.done] Building repository 'xCAT-sle15-path9' cache [....done] All repositories have been refreshed.
[c910f03c09k12]: c910f03c09k14: ospkgs: zypper --non-interactive update --auto-agree-with-licenses
c910f03c09k14: Loading repository data... Reading installed packages... Nothing to do.
[c910f03c09k12]: c910f03c09k14: ospkgs:  zypper install -y --auto-agree-with-licenses  iputils openssl rsync
c910f03c09k14: Loading repository data... Reading installed packages... 'openssl' is already installed. No update candidate for 'openssl-1.1.0h-1.11.noarch'. The highest available version is already installed. 'rsync' is already installed. No update candidate for 'rsync-3.1.3-2.10.ppc64le'. The highest available version is already installed. Resolving package dependencies... The following 2 NEW packages are going to be installed: iputils libidn11 2 new packages to install. Overall download size: 176.5 KiB. Already cached: 0 B. After the operation, additional 808.2 KiB will be used. Continue? [y/n/...? shows all options] (y): y Retrieving package libidn11-1.34-1.9.ppc64le (1/2), 75.2 KiB (345.0 KiB unpacked) Retrieving: libidn11-1.34-1.9.ppc64le.rpm [done] Retrieving package iputils-s20161105-6.10.ppc64le (2/2), 101.3 KiB (463.2 KiB unpacked) Retrieving: iputils-s20161105-6.10.ppc64le.rpm [done] Checking for file conflicts: [.......done] (1/2) Installing: libidn11-1.34-1.9.ppc64le [.........done] (2/2) Installing: iputils-s20161105-6.10.ppc64le [..........done] Additional rpm output: setting /usr/bin/ping to root:root 0755 "= cap_net_raw+ep". (wrong missing capabilities)
[c910f03c09k12]: c910f03c09k14: postscript: ospkgs exited with code 0
[c910f03c09k12]: c910f03c09k14: //xcatpost/mypostscript return with 0
[c910f03c09k12]: c910f03c09k14: Running of postscripts has completed.
```

**Comment by bybai** [comment]
7. commit bb5ffb1 for https://github.com/xcat2/xcat2-task-management/issues/296
UT link: http://10.5.106.1/install/sle15/xcat-core/docs/build/html/guides/admin-guides/manage_clusters/ppc64le/diskful/copy_image.html

File: xCAT-server/lib/xcat/plugins/sles.pm [hunk 1]
```diff
@@ -937,7 +937,11 @@ sub mkinstall
         # trim the "/" in /install/sles11.3/x86_64/
         $pkgdir =~ s/\/$//;
         if ($pkgdir =~ /^($installroot\/$os\/$arch)$/) {
-            $srcdirs[0] = "$pkgdir/1";
+            if ( -d "$pkgdir/2") {
+                $srcdirs[0] = "$pkgdir/1,$pkgdir/2";
```
Comment (immarvin) [inline xCAT-server/lib/xcat/plugins/sles.pm]: $srcdirs = ["$pkgdir/1","$pkgdir/2"]?

**Comment by bybai** [comment]
8. commit b52525f UT on x86_64:
The final repo list in compute node number should be the same in MN Package iso content:
1. in MN, modules and products number is 18, installer repo is 1:
```
bybc0605:/install/autoinst # ls /install/sle15/x86_64/2/
media.1                      Module-Development-Tools  Module-SAP-Applications     Product-SLED
Module-Basesystem            Module-HPC                Module-Server-Applications  Product-SLES
Module-CAP-Tools             Module-Legacy             Module-Web-Scripting        Product-SLES_SAP
Module-Containers            Module-Live-Patching      Product-HA                  Product-WE
Module-Desktop-Applications  Module-Public-Cloud       Product-HPC
```
2. in CN, repo files number for Packages iso is 18, repo file number for Installer ISO is 1
:
```
bybc0609:/etc/zypp/repos.d # ls
Basesystem-Module-15-0.repo            Public-Cloud-Module-15-0.repo         SLES15-15-0_0.repo
Containers-Module-15-0.repo            SAP-Applications-Module-15-0.repo     SLES15-15-0.repo
Desktop-Applications-Module-15-0.repo  Server-Applications-Module-15-0.repo  SLEWE15-15-0.repo
Development-Tools-Module-15-0.repo     SLE-15-HPC-15-0.repo                  SUSE-CAP-Tools-Module-15-0.repo
HPC-Module-15-0.repo                   SLE-15-SAP-15-0.repo                  Web-Scripting-Module-15-0.repo
Legacy-Module-15-0.repo                SLED15-15-0.repo
Live-Patching-15-0.repo                SLEHA15-15-0.repo
bybc0609:/etc/zypp/repos.d # ls |wc
     19      19     493
```

**Comment by immarvin** [comment]
i am ok with the change

**Comment by immarvin** [comment]
agree to merge this

**Comment by bybai** [comment]
Hi @jjohnson42,
I enhanced sle15 diskful support in this PR and made summarized here to let you know about them better.

The following 2 changes are different with your initial/diskless commits.
1. I replace `SLE_HPC` to `SLES` in xCAT-server/share/xcat/install/sles/compute.sle15.tmpl as default product. Since there is no `SLE_HPC` product in `SLE-15-Packages-ppc64le-GM-DVD1.iso`, and `SLES` is more general as default product for our user. 
2. `copycds` find templates based on osname, in order to support both sles12 and sle15 OS, create link `/opt/xcat/share/xcat/install/sle -> /opt/xcat/share/xcat/install/sles` and `/opt/xcat/share/xcat/netboot/sle -> /opt/xcat/share/xcat/netboot/sles`. I found you created new directory `/opt/xcat/share/xcat/netboot/sle` in diskless commits, but most files are duplicate in `/opt/xcat/share/xcat/netboot/sle` and `/opt/xcat/share/xcat/netboot/sles`, I will look into this part later.

More enhancements are done here:
1. support PPC64le
2. created `zypper` repos for all products and modules from sle15 Package ISO in compute node, so that user can add any package in pkglist  
3. added `iputils` and `vim` in sle15 pkglist, `ping` is used by xCAT postscripts.
4. `ospkgs` script sle15 support

**Pull request merged**